[spec] Auth: Token validation (#408)
modelcontextprotocol/modelcontextprotocol: docs/specification/draft/basic/authorization.mdx

Authorization servers **SHOULD** only automatically redirect the user agent if it trusts the redirection URI. If the URI is not trusted, the authorization server MAY inform the user and rely on the user to make the correct decision.

> ### 3.4 Token Validation

Reviewer: ```suggestion
### 3.4 Access Token Privilege Restriction
```

> An attacker can gain unauthorized access or perform token redirection attacks if an MCP server accepts tokens with incorrect audience claims. This vulnerability has two critical dimensions:

Reviewer: ```suggestion
An attacker can gain unauthorized access or otherwise compromise a MCP server if the server accepts tokens issued for other resources.

This vulnerability has two critical dimensions:
```

> 1. **Audience validation failures.** When an MCP server doesn't verify that tokens were specifically intended for it (via the `aud` claim), it may accept tokens originally issued for other services. This breaks the fundamental OAuth security boundary, allowing attackers to reuse legitimate tokens across different services than intended.

Reviewer: ```suggestion
1. **Audience validation failures.** When an MCP server doesn't verify that tokens were specifically intended for it (for example, via the audience claim, as mentioned in [RFC9068](https://www.rfc-editor.org/rfc/rfc9068.html)), it may accept tokens originally issued for other services. This breaks the fundamental OAuth security boundary, allowing attackers to reuse legitimate tokens across different services than intended.
```

  1. Enable privilege escalation across service boundaries
  1. Facilitate lateral movement through connected resources
  1. Allow circumvention of security controls between services
  1. Enable replay attacks against multiple backend systems

MCP servers **MUST** take all necessary steps to ensure no data is returned to unauthorized parties and **MUST** ensure any credentials are valid before processing the request.

For example, a MCP server could validate inbound tokens through one of the following approaches:

1. Token introspection, according to [RFC7662](https://datatracker.ietf.org/doc/html/rfc7662).  1. JWT validation, according to [RFC 9068](https://www.rfc-editor.org/rfc/rfc9068.html).
1. Custom validation, according to the conventions established by the AS.

MCP servers **MUST** strictly validate token audiences and only accept tokens specifically intended for themselves. Implementers **MUST NOT** design architectures where clients send

> tokens through the MCP server intended for other resources.

Reviewer: ```suggestion
tokens through the MCP server intended for other resources.

The access token of the upstream API may be an opaque token and the MCP server has no way to introspect the token to validate the audience or know what user it is associated with. It **MUST NOT** accept such tokens to grant access to its resources.
```

2. **Token passthrough.** If the MCP server not only accepts tokens with incorrect audiences but also forwards these unmodified tokens to downstream services, it introduces the "confused deputy" risk, outlined in [Section 3.4](#34-confused-deputy-problem). The MCP server becomes an unintentional proxy that can:
  1. Enable privilege escalation across service boundaries
  1. Facilitate lateral movement through connected resources
  1. Allow circumvention of security controls between services
  1. Enable replay attacks against multiple backend systems

MCP servers **MUST** take all necessary steps to ensure no data is returned to unauthorized parties and **MUST** ensure any credentials are valid before processing the request.

For example, a MCP server could validate inbound tokens through one of the following approaches:

> 1. Token introspection, according to [RFC7662](https://datatracker.ietf.org/doc/html/rfc7662).  1. JWT validation, according to [RFC 9068](https://www.rfc-editor.org/rfc/rfc9068.html).

Reviewer 2: ```suggestion
1. Token introspection, according to [RFC7662](https://datatracker.ietf.org/doc/html/rfc7662).  
1. JWT validation, according to [RFC 9068](https://www.rfc-editor.org/rfc/rfc9068.html).
```

1. Custom validation, according to the conventions established by the AS.

MCP servers **MUST** strictly validate token audiences and only accept tokens specifically intended for themselves. Implementers **MUST NOT** design architectures where clients send
tokens through the MCP server intended for other resources.

> The access token of the upstream API may be an opaque token and the MCP server has no way to introspect the token to validate the audience or know what user it is associated with. It **MUST NOT** accept such tokens to grant access to its resources.

Reviewer 2: ```suggestion
If the MCP server makes requests to upstream APIs, it acts as an OAuth client to the upstream API. The access token it uses at the upstream API may be an opaque token, so the MCP server has no way to introspect the token to validate the audience or know what user it is associated with. Even if the token were a JWT [RFC 9068](https://www.rfc-editor.org/rfc/rfc9068.html) token, the audience of that token would not be the MCP server, so the MCP server is not intended to parse the token anyway. The MCP server **MUST NOT** accept such tokens to grant access to its resources.
```

MCP proxy servers using static client IDs **MUST** obtain user consent for each dynamically
registered client before forwarding to third-party authorization servers (which may require additional consent).

### 3.5 Access Token Privilege Restriction

An attacker can gain unauthorized access or otherwise compromise a MCP server if the server accepts tokens issued for other resources.

This vulnerability has two critical dimensions:

> 1. **Audience validation failures.** When an MCP server doesn't verify that tokens were specifically intended for it (for example, via the audience claim, as mentioned in [RFC9068](https://www.rfc-editor.org/rfc/rfc9068.html)), it may accept tokens originally issued for other services. This breaks the fundamental OAuth security boundary, allowing attackers to reuse legitimate tokens across different services than intended.

Reviewer 2: ```suggestion
1. **Audience validation failures.** When an MCP server doesn't verify that tokens were specifically intended for it (for example, via the audience claim, as mentioned in [RFC9068](https://www.rfc-editor.org/rfc/rfc9068.html)), it may accept tokens originally issued for other services. This breaks a fundamental OAuth security boundary, allowing attackers to reuse legitimate tokens across different services than intended.
```

2. **Token passthrough.** If the MCP server not only accepts tokens with incorrect audiences but also forwards these unmodified tokens to downstream services, it introduces the "confused deputy" risk, outlined in [Section 3.4](#34-confused-deputy-problem). The MCP server becomes an unintentional proxy that can:
  1. Enable privilege escalation across service boundaries
  1. Facilitate lateral movement through connected resources
  1. Allow circumvention of security controls between services

>   1. Enable replay attacks against multiple backend systems

Reviewer 3: I've been reviewing this section and I'm wondering if we could enhance the clarity of the risk description to ensure it resonates with all readers, particularly implementors who may push for clarification.

For instance, I noticed that the opening reference to token passthrough points to Section 3.4, but I found myself having trouble connecting those specific scenarios since Section 3.4 and the related Best Practices document seems to address a different security context that doesn't rely on token passthrough. Would it be helpful to either adjust the reference or expand on how these scenarios interconnect? 

Regarding the four numbered impacts, I think adding concrete examples (similar to how we illustrated the confused deputy problem) would make the risks more tangible and actionable for readers, perhaps moving the details to the Security Best Practices document if necessary for space reasons. For instance, with points like "Facilitate lateral movement through connected resources" and "replay attacks against multiple backend systems", I wasn't quite sure how token passthrough enables them. 

I was considering how I would explain the risks of token passthrough to peers. These are the topics that came to mind.
1. Security Control Circumvention:
   - The MCP Server might be implementing important security controls like rate limiting, request validation, or traffic monitoring. If clients can obtain and use tokens directly, they bypass these controls
2. Accountability and Audit Trail Issues:
   - The MCP Server will be unable to identify or distinguish between MCP Clients when clients are calling with an upstream-issued access token which may be opaque to the MCP Server. 
   - The upstream Resource Server's logs will show requests coming from the MCP Server's client ID, but the actual requests are originating from different clients If clients obtain and use tokens directly.
   - Both factors make incident investigation and auditing more difficult.
3. Trust Boundary Issues:
   - The upstream Resource Server granted trust to the MCP Server, not to arbitrary clients. This trust might include assumptions about origin or client behavior patterns. Breaking this trust boundary could lead to unexpected issues.
4. Future Compatibility Risk:
   - Even if an MCP Server starts as a "pure proxy" today, it might need to add security controls later. Starting with proper token audience separation makes it easier to evolve the security model.

The other option for expediency is to remove the detailed explanations entirely for now, just refer to the OAuth requirements, and follow-up later with a seperate update to Best Practices that explains the risks in details, if we see demand for it.

Reviewer 4: Hello! I want to voice my support of the contents of this PR - I think it mentions important risks and considerations often overlooked by people new to the involved OAuth spec. That said, I think this is a valid point about having the recommendations resonate easily. My two cents:

- **Token validation with audience check** by itself is important for communication between the MCP Client and MCP Server, regardless of whether token passthrough is happening. This is essential to securing specifically the MCP Server against token misuse, is a relatively simple recommendation. Therefore, I think it belongs in this main Authorization doc
- The various **risks of token passthrough** involve moreso the communication between the MCP Server and the underlying tool and the securing of said tool. While certainly related and important, I do wonder if it's outside the scope of this main Authorization doc and should be placed in some separate doc either more related to either a supplementary "security best practices" doc that can be mentioned from the "MCP Server" part of the spec. I also wonder if the related recommendations may be more involved for many different tools (potentially involving token exchange alluded to in #214, or what if the underlying tool is secured by a different Auth server, not OAuth2.0-secured, etc.)

> MCP servers **MUST** take all necessary steps to ensure no data is returned to unauthorized parties and **MUST** ensure any credentials are valid before processing the request.

Reviewer 2: ```suggestion
MCP servers **MUST** validate access tokens before processing the request, and take all necessary steps to ensure no data is returned to unauthorized parties.
```

Are there any credentials expected other than an access token? Seems better to be explicit here.

> A MCP server **CAN** follow the guidelines in [OAuth 2.1 - Section 5.2](https://www.ietf.org/archive/id/draft-ietf-oauth-v2-1-12.html#section-5.2) to validate inbound tokens.

Reviewer 2: Section 5.2 is a generic "you MUST validate access tokens" section, and points to two RFCs for ways to accomplish this, but the first part of 5.2 is a MUST, so it seems weird to say "CAN" here. (Also "CAN" is not an RFC2119 keyword.) Ultimately the authorization server is the one that decides the token format, so the MCP server doesn't get to make much of a decision on *how* it validates tokens, it will be doing whatever the authorization server says to do. I'm struggling to find any actionable guidance for MCP servers in this sentence, might be best to just take it out entirely.

MCP servers **MUST** only accept tokens specifically intended for themselves.

If the MCP server makes requests to upstream APIs, it may act as an OAuth client to them. The access token used at the upstream API is a seperate token, issued by the upstream authorization server. The MCP server **MUST NOT** pass through the token it received from the MCP client.

> Additional safeguards may be implemented to protect against token misuse, as defined in [RFC 8707](https://www.rfc-editor.org/rfc/rfc8707.html).

Reviewer 2: RFC 8707 defines a new parameter in the authorization request, so that's a bit beyond the scope of "additional safeguards" IMO. Is the intent to suggest using the resource parameter as described in the spec?

Reviewer: Updated the verbiage.

> If the authorization server supports the `resource` parameter, it's is recommended that implementers follow [RFC 8707](https://www.rfc-editor.org/rfc/rfc8707.html) to prevent token misuse.

Reviewer 3: Typo: "it's is"

  1. Enable privilege escalation across service boundaries
  1. Facilitate lateral movement through connected resources
  1. Allow circumvention of security controls between services
  1. Enable replay attacks against multiple backend systems

MCP servers **MUST** take all necessary steps to ensure no data is returned to unauthorized parties and **MUST** ensure any credentials are valid before processing the request.

For example, a MCP server could validate inbound tokens through one of the following approaches:

1. Token introspection, according to [RFC7662](https://datatracker.ietf.org/doc/html/rfc7662).  
1. JWT validation, according to [RFC 9068](https://www.rfc-editor.org/rfc/rfc9068.html).
1. Custom validation, according to the conventions established by the AS.

MCP servers **MUST** strictly validate token audiences and only accept tokens specifically intended for themselves. Implementers **MUST NOT** design architectures where clients send
tokens through the MCP server intended for other resources.

> If the MCP server makes requests to upstream APIs, it acts as an OAuth client to the upstream API. The access token it uses at the upstream API may be an opaque token, so the MCP server has no way to introspect the token to validate the audience or know what user it is associated with. Even if the token were a JWT [RFC 9068](https://www.rfc-editor.org/rfc/rfc9068.html) token, the audience of that token would not be the MCP server, so the MCP server is not intended to parse the token anyway. The MCP server **MUST NOT** accept such tokens to grant access to its resources.

Reviewer 3: MCP servers **MUST** only accept tokens specifically intended for themselves, as outlined in [OAuth 2.1 section 5.1](https://www.ietf.org/archive/id/draft-ietf-oauth-v2-1-12.html#section-5.2). 

If the MCP server makes requests to upstream APIs, it may act as an OAuth client to the upstream API. The access token used at the upstream API is a seperate token, issued by the upstream authorization server. The MCP server **MUST NOT** pass through the token it received from the MCP client.

Reviewer 5: While agreeing to the suggested token validation at the MCP Server, when the upstream API serves resources that are intended for the resource owner, how does the MCP server is expected to obtain a resource owner delegated access token.

Reviewer: @ayshsandu that is up to the implementer, but it might be that your question is somewhat orthogonal to the broader token validation question. If you are gating your server behind a specific AS, that AS _must_ be able to issue tokens that are destined _for that server_.

Reviewer 5: @localden I'm trying to understand how the MCP server obtain a user delegated token from the AS at this point. As there will be a higher demand to expose existing APIs as MCP servers, some guideline on this would be benificial.
![mermaid-diagram-2025-05-05-145802](https://github.com/user-attachments/assets/d1f72da2-6b4e-4b9c-b044-895db76989fc)
